File: r3-calculator/calculator.py
import sys


DIV = '/'
PLUS = '+'
MULT = '*'
MINUS = '-'
NUM = 'int'
OPEN_COMMENT = '{'
CLOSE_COMMENT = '}'
STD_FILE_NAME = 'test.in'


class Token:
    operators = []

    def __init__(self, type_, value):
        self.type_ = type_
        self.value = value

class Term(Token):
    operators = [MULT, DIV]

class Num(Token):
    operators = []

class Expr(Token):
    operators = [PLUS, MINUS]

class Tokenizer:
    def __init__(self, src, pos=0, curr=None):
        self.src = src
        self.pos = pos
        self.curr = curr
        self.is_comment = False

    def get_next(self):
        if self.pos < len(self.src):
            self._read()
            return self.curr
        else:
            return None

    def _read(self):
        self.curr = self._read_any()

    def _read_any(self):
        curr_token = self.src[self.pos]
        if not self.is_comment:
            if curr_token in Term.operators:
                return self._read_term()
            elif curr_token in Expr.operators:
                return self._read_operator()
            elif curr_token.isdigit():
                return self._read_int()
            elif curr_token.isspace():
                self.pos += 1
                return self._read_any()
            elif curr_token == OPEN_COMMENT:
                self.pos += 1
                self.is_comment = True
                return self._read_any()
            else:
                raise ValueError('Unexpected token at index {id_}: {token}'
                                 .format(id_=self.pos,
                                         token=self.src[self.pos]))
        elif curr_token == CLOSE_COMMENT:
            self.pos += 1
            self.is_comment = False
            return self._read_any()
        else:
            self.pos += 1
            return self._read_any()

    def _read_term(self):
        curr_token = self.src[self.pos]
        self.pos += 1
        if curr_token == MULT:
            return Term(MULT, None)
        elif curr_token == DIV:
            return Term(DIV, None)
        else:
            self.pos -= 1
            raise ValueError('Unexpected token at index {id_}: {token}'
                             .format(id_=self.pos,
                                     token=self.src[self.pos]))

    def _read_operator(self):
        current_token = self.src[self.pos]
        self.pos +=1
        if current_token == '+':
            return Expr(PLUS, None)
        elif current_token == '-':
            return Expr(MINUS, None)
        else:
            self.pos -= 1
            raise ValueError('Unexpected token at index {id_}: {token}'
                             .format(id_=self.pos,
                                     token=self.src[self.pos]))

    def _read_int(self):
        # actually reads an expression with * or / until it "wraps" in a + or -
        # token
        if not self.src[self.pos].isdigit():
            raise ValueError('Unexpected token at index {id_}: {token}'
                             .format(id_=self.pos,
                                     token=self.src[self.pos]))
        val = 0
        while self.src[self.pos].isdigit():
            val = val * 10 + int(self.src[self.pos])
            self.pos += 1
            if self.pos >= len(self.src):
                break

        return Num(NUM, val)

class Parser:
    def __init__(self, src):
        self.tokens = Tokenizer(src)

    def analyse_term(self):
        self.val = self.tokens.get_next()
        res = self.val.value
        self.val = self.tokens.get_next()

        while self.val is not None and self.val.type_ in Term.operators:
            if self.val.type_ == MULT:
                self.val = self.tokens.get_next()
                if self.val.type_ == NUM:
                    res *= self.val.value
                else:
                    raise ValueError('Unexpected token type, expected int, got {}'
                                     .format(self.val.type_))

            elif self.val.type_ == DIV:
                self.val = self.tokens.get_next()
                if self.val.type_ == NUM:
                    res //= self.val.value
                else:
                    raise ValueError('Unexpected token type, expected int, got {}'
                                     .format(self.val.type_))

            self.val = self.tokens.get_next()
        return res

    def analyse_exp(self):
        res = self.analyse_term()
        while self.val is not None:
            if self.val.type_ == PLUS:
                res += self.analyse_term()
            elif self.val.type_ == MINUS:
                res -= self.analyse_term()
        return res

if __name__ == '__main__':
    # for debugging
    try:
        file_name = sys.argv[1]
    except IndexError:
        # no arg was passed
        file_name = STD_FILE_NAME

    try:
        with open(file_name, 'r') as fin:
            for line in fin:
                line = line.strip()
                parser = Parser(line)
                print('{} = {}'.format(line, parser.analyse_exp()))
    except IOError as err:
        print(err)

File: r3-calculator/calculator.go
package main

// [WIP] (!!!!)

import (
	"os"
	"fmt"
	"log"
	"bufio"
	"errors"
	"unicode"
)

const DIV = '/'
const PLUS = '+'
const MULT = '*'
const MINUS = '-'
const NUM = "int"
const OPEN_PARENT = '('
const CLOSE_PARENT = ')'
const OPEN_COMMENT = '{'
const CLOSE_COMMENT = '}'
const STD_FILE_NAME = "test.in"
var NUMS_OP = map[rune]bool{}
var TERM_OP = map[rune]bool {MULT: true, DIV: true}
var EXPR_OP = map[rune]bool {PLUS: true, MINUS: true}  // same for sign_op
var PARENT_OP = map[rune]bool {OPEN_PARENT: true, CLOSE_PARENT: true}

type Token struct {
	value int
	type_ string
}

type Tokenizer struct {
	is_comment bool
	pos int
	src string
	curr *Token
}

type Parser struct {
	tokens *Tokenizer
}

func (t *Tokenizer) GetNext() (*Token, error) {
	if t.pos < len(t.src) {
		if err := t.read(); err != nil {
			return nil, err
		}
		return t.curr, nil
	} else {
		return nil, nil
	}
}

func (t *Tokenizer) read() error {
	// tbh i dont know why i'm using both read and readany at this point...
	var err error
	t.curr, err = t.readAny()
	return err
}

func (t *Tokenizer) readAny() (*Token, error) {
	curr_token := rune(t.src[t.pos])
	if !t.is_comment {
		if _, present := TERM_OP[curr_token]; present {
			return t.readTerm(curr_token)
		} else if _, present := EXPR_OP[curr_token]; present {
			return t.readExpr(curr_token)
		} else if _, present := PARENT_OP[curr_token]; present {
			return t.readParent(curr_token)
		} else if unicode.IsDigit(curr_token) {
			return t.readInt()
		} else if unicode.IsSpace(curr_token) {
			t.pos++
			return t.readAny()
		} else if curr_token == OPEN_COMMENT {
			t.pos++
			t.is_comment = true
			return t.readAny()
		} else {
			return nil, errors.New(
				fmt.Sprintf("Unexpected token at index %d: %s", t.pos, curr_token))
		}
	} else if curr_token == CLOSE_COMMENT {
		t.pos++
		t.is_comment = false
		return t.readAny()
	} else {
		t.pos++
		return t.readAny()
	}
	return nil, errors.New("How did you even get here?")
}

func (t *Tokenizer) readParent(curr_token rune) (*Token, error) {
	t.pos++  // prolly should be incrementing position in #readAny
	return &Token{0, string(curr_token)}, nil
}

func (t *Tokenizer) readTerm(curr_token rune) (*Token, error) {
	t.pos++  // this is literally the same func as above... what have i done ;-;
	return &Token{0, string(curr_token)}, nil
}

func (t *Tokenizer) readExpr(curr_token rune) (*Token, error) {
	t.pos++ // oh not again...
	return &Token{0, string(curr_token)}, nil
}

func (t *Tokenizer) readInt() (*Token, error) {
	// actually reads an expression with * or / until it "wraps" in a + or - token
	val := 0
	for t.pos < len(t.src) && unicode.IsDigit(rune(t.src[t.pos])) {
		val = val * 10 + rtoi(rune(t.src[t.pos]))
		t.pos++
	}
	return &Token{val, NUM}, nil
}

func rtoi(r rune) int {
	return int(r) - '0'
}

func validInMap(token *Token, test_map map[rune]bool) bool {
	if token != nil {
		if _, present := test_map[rune(token.type_[0])]; present {
			return true
		}
	}
	return false
}

func InitParser(src string) Parser {
	// """""constructor""""""
	return Parser{&Tokenizer{false, 0, src, nil}}
}

func (p Parser) AnalyseExp() (int, error) {
	res := 0
	var err error
	res, err = p.analyseTerm()
	if err != nil {
		return 0, err
	}
	for validInMap(p.tokens.curr, EXPR_OP) {
		currType := rune(p.tokens.curr.type_[0])
		if currType == PLUS {
			if val, err := p.analyseTerm(); err != nil {
				res += val
			}
		} else if currType == MINUS {
			if val, err := p.analyseTerm(); err != nil {
				res -= val
			}
		}
		if err != nil {
			return 0, err
		}
	}
	return res, nil
}

func (p Parser) analyseTerm() (int, error) {
	res := 0
	var err error
	if res, err = p.analyseFactor(); err != nil {
		return 0, err
	}
	if _, err := p.tokens.GetNext(); err != nil {
		return res, err
	}
	for validInMap(p.tokens.curr, TERM_OP) {
		currType := rune(p.tokens.curr.type_[0])
		if currType == MULT {
			if val, err := p.analyseFactor(); err != nil {
				res *= val
			}
		} else if currType == DIV {
			if val, err := p.analyseFactor(); err != nil {
				res /= val
			}
		}
		if err != nil {
			return res, err
		}
		if _, err = p.tokens.GetNext(); err != nil {
			return res, err
		}
	}
	return res, nil
}

func (p Parser) analyseFactor() (int, error) {
	var err error
	if _, err = p.tokens.GetNext(); err != nil {
		return 0, err
	}
	currType := rune(p.tokens.curr.type_[0])
	if p.tokens.curr.type_ == NUM {
		return p.tokens.curr.value, nil
	} else if currType == OPEN_PARENT {
		return p.analyseParent()
	} else if currType == MINUS {
		return p.analyseUnary(MINUS)
	} else if currType == PLUS {
		return p.analyseUnary(PLUS)
	} else {
		return 0, errors.New(
			fmt.Sprintf("Unexpected token type, expected factor token, got %s",
						p.tokens.curr.type_))
	}
	return 0, errors.New("How did you even get here?!")
}

func (p Parser) analyseUnary(sign rune) (int, error) {
	if sign == PLUS {
		return p.analyseFactor()
	} else if sign == MINUS {
		res, err := p.analyseFactor()
		return -res, err
	}
	return 0, errors.New(
		fmt.Sprintf("Unexpected token type, expected sign (+ or -) token, got %s", sign))
}

func (p Parser) analyseParent() (int, error) {
	res, err := p.AnalyseExp()
	if err != nil {
		return 0, err
	}
	currType := rune(p.tokens.curr.type_[0])
	if currType != CLOSE_PARENT {
		return 0, errors.New(
			fmt.Sprintf("Unexpected token type, expected ')', got %s", currType))
	}
	return res, nil
}

func main() {
	file, err := os.Open("./test.in")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()   // guarantees this will happen at the end!
	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)

	for scanner.Scan() {
		line := scanner.Text()
		parser := InitParser(line)
		res, err := parser.AnalyseExp()
		if err != nil {
			log.Fatal(err)
			return
		}
		fmt.Printf("%s = %d\n", line, res)
	}
}